crates/pep508-rs/src/marker/algebra.rs

```diff
@@ -748,6 +751,110 @@ impl Edges {
     }
 }
 
+// Normalize a [`VersionSpecifier`] before adding it to the tree.
+fn normalize_specifier(specifier: VersionSpecifier) -> VersionSpecifier {
+    let (operator, version) = specifier.into_parts();
+
+    // The decision diagram relies on the assumption that the negation of a marker tree is
+    // the complement of the marker space. However, pre-release versions violate this assumption.
+    // For example, the marker `python_full_version > '3.9' or python_full_version <= '3.9'`
+    // does not match `python_full_version == 3.9.0a0`. However, it's negation,
+    // `python_full_version > '3.9' and python_full_version <= '3.9'` also does not include
+    // `3.9.0a0`, and is actually `false`.
+    //
+    // For this reason we ignore pre-release versions entirely when evaluating markers.
+    // Note that `python_version` cannot take on pre-release values so this is necessary for
+    // simplifying ranges, but for `python_full_version` this decision is a semantic change.
+    let mut release = version.release();
+
+    // Strip any trailing `0`s.
+    //
+    // The [`Version`] type ignores trailing `0`s for equality, but still preserves them in it's
+    // [`Display`] output. We must normalize all versions by stripping trailing `0`s to remove the
+    // distinction between versions like `3.9` and `3.9.0`, whose output will depend on which form
+    // was added to the global marker interner first.
+    //
+    // Note that we cannot strip trailing `0`s for star equality, as `==3.0.*` is different from `==3.*`.
+    if !operator.is_star() {
+        if let Some(end) = release.iter().rposition(|segment| *segment != 0) {
+            if end > 0 {
+                release = &release[..=end];
+            }
+        }
+    }
+
+    VersionSpecifier::from_version(operator, Version::new(release)).unwrap()
+}
+
+/// Returns the equivalent `python_full_version` specifier for a `python_version` comparison.
+///
+/// Returns `Err` with a constant node if the equivalent comparison is always `true` or `false`.
+fn python_version_to_full_version(specifier: VersionSpecifier) -> Result<VersionSpecifier, NodeId> {
+    let major_minor = match *specifier.version().release() {
+        // `python_version == 3.*` is equivalent to `python_full_version == 3.*`
+        // and adding a trailing `0` would be incorrect.
+        [_major] if specifier.operator().is_star() => return Ok(specifier),
+        // Note that `python_version == 3` matches `3.0.1`, `3.0.2`, etc.
+        [major] => Some((major, 0)),
+        [major, minor] => Some((major, minor)),
+        _ => None,
```

Comment:
Maybe it'd be helpful to suggest what we'll do in this case?
```suggestion
        // Specifiers including segments beyond the minor version require separate handling
        _ => None,
```